gui/widgets/groupedlist.cpp

```diff
@@ -0,0 +1,427 @@
+			int groupID = indexToGroupID(_listIndex[newSelectedItem]);
+			_selectedItem = -1;
+			toggleGroup(groupID);
+			warning("%d", groupID);
+		}
+	}
+
+	// TODO: Determine where inside the string the user clicked and place the
+	// caret accordingly.
+	// See _editScrollOffset and EditTextWidget::handleMouseDown.
+	markAsDirty();
+
+}
+
+void GroupedListWidget::handleMouseUp(int x, int y, int button, int clickCount) {
+	// If this was a double click and the mouse is still over
+	// the selected item, send the double click command
+	if (clickCount == 2 && (_selectedItem == findItem(x, y))) {
+		int selectID = getSelected();
+		if (selectID >= 0) {
+			sendCommand(kListItemDoubleClickedCmd, _selectedItem);
+		}
+	}
+}
+
+void GroupedListWidget::handleMouseWheel(int x, int y, int direction) {
+	_scrollBar->handleMouseWheel(x, y, direction);
+}
+
+void GroupedListWidget::handleCommand(CommandSender *sender, uint32 cmd, uint32 data) {
+	switch (cmd) {
+	case kSetPositionCmd:
+		if (_currentPos != (int)data) {
+			_currentPos = data;
+			checkBounds();
+			markAsDirty();
+
+			// Scrollbar actions cause list focus (which triggers a redraw)
+			// NOTE: ListWidget's boss is always GUI::Dialog
+			((GUI::Dialog *)_boss)->setFocusWidget(this);
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+void GroupedListWidget::reflowLayout() {
+	Widget::reflowLayout();
+
+	_leftPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Left", 0);
+	_rightPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Right", 0);
+	_topPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Top", 0);
+	_bottomPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Bottom", 0);
+	_hlLeftPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.hlLeftPadding", 0);
+	_hlRightPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.hlRightPadding", 0);
+
+	_scrollBarWidth = g_gui.xmlEval()->getVar("Globals.Scrollbar.Width", 0);
+
+	// HACK: Once we take padding into account, there are times where
+	// integer rounding leaves a big chunk of white space in the bottom
+	// of the list.
+	// We do a rough rounding on the decimal places of Entries Per Page,
+	// to add another entry even if it goes a tad over the padding.
+	frac_t entriesPerPage = intToFrac(_h - _topPadding - _bottomPadding) / kLineHeight;
+
+	// Our threshold before we add another entry is 0.9375 (0xF000 with FRAC_BITS being 16).
+	const frac_t threshold = intToFrac(15) / 16;
+
+	if ((frac_t)(entriesPerPage & FRAC_LO_MASK) >= threshold)
+		entriesPerPage += FRAC_ONE;
+
+	_entriesPerPage = fracToInt(entriesPerPage);
+	assert(_entriesPerPage > 0);
+
+	if (_scrollBar) {
+		_scrollBar->resize(_w - _scrollBarWidth, 0, _scrollBarWidth, _h, false);
+		scrollBarRecalc();
+		scrollToCurrent();
+	}
+}
+
+void GroupedListWidget::toggleGroup(int groupID) {
+	// Shrink group if it is expanded
+	if (_groupExpanded[groupID]) {
+		_groupExpanded[groupID] = false;
```

Comment:
This can also be written as simply

 _groupExpanded[groupId] = !_groupExpanded[groupId]
 sortGroups()
 
 Which is perhaps slightly less readable, so a matter of taste.

Reply:
I think I was earlier performing checks to see if the selection went out of bounds on shrinking a group so that I could move the selection to something else, so I made the two branches. However, it didn't work out too well, so now I just set the selection to -1, whenever a group header is interacted with. I may try to figure out the earlier mentioned problem later though, so I will keep the branches as is.

Reply:
I have not been able to make use of the branches. I have replaced the branching in 32eafca, it may be branched later on if required.